File: IDHAR_UDHAR_ADMIN/src/services/fareEngine.js
export function round2(value) {
  return Math.round((Number(value) + Number.EPSILON) * 100) / 100;
}

export const DEFAULT_FARE_BY_CATEGORY = {
  Bike: { baseFare: 79, perKmCharge: 0, initialMinimum: 79, waitingCharge: 0, surgeCharge: 0, tollCharge: 0, parkingCharge: 0, weightCapacityKg: 20, size: '36cm' },
  Auto: { baseFare: 149, perKmCharge: 0, initialMinimum: 149, waitingCharge: 0, surgeCharge: 0, tollCharge: 0, parkingCharge: 0, weightCapacityKg: 100, size: '' },
  'Mini Truck': { baseFare: 399, perKmCharge: 0, initialMinimum: 399, waitingCharge: 0, surgeCharge: 0, tollCharge: 0, parkingCharge: 0, weightCapacityKg: 500, size: '' },
  Tempo: { baseFare: 499, perKmCharge: 0, initialMinimum: 499, waitingCharge: 0, surgeCharge: 0, tollCharge: 0, parkingCharge: 0, weightCapacityKg: 900, size: '' },
  'Large Tempo': { baseFare: 599, perKmCharge: 0, initialMinimum: 599, waitingCharge: 0, surgeCharge: 0, tollCharge: 0, parkingCharge: 0, weightCapacityKg: 1500, size: '' },
  Truck: { baseFare: 699, perKmCharge: 0, initialMinimum: 699, waitingCharge: 0, surgeCharge: 0, tollCharge: 0, parkingCharge: 0, weightCapacityKg: 1000, size: '' },
};

export function emptyFareFields() {
  return {
    baseFare: 0,
    perKmCharge: 0,
    initialMinimum: 0,
    waitingCharge: 0,
    surgeCharge: 0,
    tollCharge: 0,
    parkingCharge: 0,
    weightCapacityKg: '',
    size: '',
    fareVersionId: 'fare_v1',
  };
}

/** GST is not applied. */
export function quoteFare(config, distanceKm = 0, discount = 0) {
  const baseFare = Number(config.baseFare || 0);
  const perKmCharge = Number(config.perKmCharge || 0);
  const initialMinimum = Number(config.initialMinimum || 0);
  const waitingCharge = Number(config.waitingCharge || 0);
  const surgeCharge = Number(config.surgeCharge || 0);
  const tollCharge = Number(config.tollCharge || 0);
  const parkingCharge = Number(config.parkingCharge || 0);
  const distanceCharge = round2(perKmCharge * Number(distanceKm || 0));
  const raw = baseFare + distanceCharge + waitingCharge + surgeCharge + tollCharge + parkingCharge;
  const tripFare = raw < initialMinimum ? initialMinimum : round2(raw);
  const afterDiscount = Math.max(0, tripFare - Number(discount || 0));
  const netTotal = round2(afterDiscount);
  return {
    configVersionId: config.fareVersionId || config.versionId || 'fare_v1',
    distanceKm: Number(distanceKm || 0),
    baseFare,
    perKmCharge,
    distanceCharge,
    initialMinimum,
    waitingCharge,
    surgeCharge,
    tollCharge,
    parkingCharge,
    tripFare,
    discount: Number(discount || 0),
    subtotal: netTotal,
    rounding: 0,
    tax: 0,
    netTotal,
    quotedAt: new Date().toISOString(),
  };
}

const fareHistory = [];

export function publishFareVersion(categoryId, config) {
  const version = {
    id: `fare_${categoryId}_${Date.now()}`,
    categoryId,
    ...config,
    effectiveFrom: new Date().toISOString(),
  };
  fareHistory.push(version);
  return version;
}

export function fareHistoryFor(categoryId) {
  return fareHistory.filter((row) => row.categoryId === categoryId);
}

File: idhar_udhar/lib/shared/business/fare_engine.dart
import 'fare_config.dart';

/// Quote produced at booking confirmation. Copied onto the order as a snapshot.
/// GST is not applied (NO GST).
class FareQuote {
  const FareQuote({
    required this.configVersionId,
    required this.vehicleCategoryId,
    required this.distanceKm,
    required this.baseFare,
    required this.perKmCharge,
    required this.distanceCharge,
    required this.initialMinimum,
    required this.waitingCharge,
    required this.surgeCharge,
    required this.tollCharge,
    required this.parkingCharge,
    required this.tripFare,
    required this.discount,
    required this.subtotal,
    required this.rounding,
    required this.netTotal,
    required this.quotedAt,
    this.stopCount = 1,
  });

  final String configVersionId;
  final String vehicleCategoryId;
  final double distanceKm;
  final double baseFare;
  final double perKmCharge;
  final double distanceCharge;
  final double initialMinimum;
  final double waitingCharge;
  final double surgeCharge;
  final double tollCharge;
  final double parkingCharge;
  final double tripFare;
  final double discount;
  final double subtotal;
  final double rounding;
  final double netTotal;
  final DateTime quotedAt;
  final int stopCount;

  /// Always zero. Dummy 5% tax must not return.
  double get tax => 0;

  Map<String, double> get visibleLines {
    return <String, double>{
      'Base Fare': baseFare,
      if (distanceCharge > 0) 'Distance': distanceCharge,
      if (waitingCharge > 0) 'Waiting': waitingCharge,
      if (surgeCharge > 0) 'Surge': surgeCharge,
      if (tollCharge > 0) 'Toll': tollCharge,
      if (parkingCharge > 0) 'Parking': parkingCharge,
      if (discount > 0) 'Discount': -discount,
    };
  }
}

abstract final class FareEngine {
  static double round2(double value) =>
      (value * 100).roundToDouble() / 100;

  /// Multi-stop: [distanceKm] is the sum of ordered legs.
  /// No extra multi-stop fee is invented.
  static FareQuote quote({
    required FareConfig config,
    required double distanceKm,
    int stopCount = 1,
    double discount = 0,
    DateTime? quotedAt,
  }) {
    final double distanceCharge = round2(config.perKmCharge * distanceKm);
    final double raw = config.baseFare +
        distanceCharge +
        config.waitingCharge +
        config.surgeCharge +
        config.tollCharge +
        config.parkingCharge;
    final double tripFare =
        raw < config.initialMinimum ? config.initialMinimum : round2(raw);
    final double afterDiscount = (tripFare - discount).clamp(0, double.infinity);
    final double net = round2(afterDiscount);
    final double rounding = round2(net - afterDiscount);
    return FareQuote(
      configVersionId: config.versionId,
      vehicleCategoryId: config.vehicleCategoryId,
      distanceKm: distanceKm,
      baseFare: config.baseFare,
      perKmCharge: config.perKmCharge,
      distanceCharge: distanceCharge,
      initialMinimum: config.initialMinimum,
      waitingCharge: config.waitingCharge,
      surgeCharge: config.surgeCharge,
      tollCharge: config.tollCharge,
      parkingCharge: config.parkingCharge,
      tripFare: tripFare,
      discount: discount,
      subtotal: round2(afterDiscount),
      rounding: rounding,
      netTotal: net,
      quotedAt: quotedAt ?? DateTime.now(),
      stopCount: stopCount,
    );
  }
}
